File: solo-leveling-ui-service/src/main/java/com/sleepkqq/sololeveling/ui/broadcast/AbstractTopic.java
package com.sleepkqq.sololeveling.ui.broadcast;

import com.sleepkqq.sololeveling.avro.notification.Notification;
import com.vaadin.flow.shared.Registration;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;

public abstract class AbstractTopic<T> implements Topic<T> {

  private final Map<Long, T> listeners = new ConcurrentHashMap<>();

  @Override
  public Registration subscribe(long userId, T listener) {
    listeners.put(userId, listener);
    return () -> listeners.remove(userId);
  }

  @Override
  public void publish(long userId, Notification notification) {
    Optional.ofNullable(listeners.get(userId))
        .ifPresent(l -> notifyListener(l, notification));
  }

  protected abstract void notifyListener(T listener, Notification notification);
}

File: solo-leveling-ui-service/src/main/kotlin/com.sleepkqq.sololeveling/ui/broadcast/AbstractTopic.kt
package com.sleepkqq.sololeveling.ui.broadcast

import com.sleepkqq.sololeveling.avro.notification.Notification
import com.vaadin.flow.shared.Registration
import java.util.concurrent.ConcurrentHashMap

abstract class AbstractTopic<T> : Topic<T> {

	private val listeners: MutableMap<Long, T> = ConcurrentHashMap()

	override fun subscribe(userId: Long, listener: T): Registration {
		listeners[userId] = listener
		return Registration { listeners.remove(userId) }
	}

	override fun publish(userId: Long, notification: Notification) {
		listeners[userId]?.let { notifyListener(it, notification) }
	}

	protected abstract fun notifyListener(listener: T, notification: Notification)
}
